native/cocos/renderer/pipeline/custom/NativeBuiltinUtils.cpp

```diff
@@ -873,6 +850,50 @@ const BuiltinCascadedShadowMap *getBuiltinShadowCSM(
     return &csm;
 }
 
+} // namespace
+
+const BuiltinCascadedShadowMap *getBuiltinShadowCSM(
+    const PipelineRuntime &pplRuntime,
+    const scene::Camera &camera,
+    const scene::DirectionalLight *mainLight) {
+    const auto &ppl = dynamic_cast<const NativePipeline &>(pplRuntime);
+    // no main light
+    if (!mainLight) {
+        return nullptr;
+    }
+    // not attached to a node
+    if (!mainLight->getNode()) {
+        return nullptr;
+    }
+    const pipeline::PipelineSceneData &pplSceneData = *pplRuntime.getPipelineSceneData();
+    auto &csmLayers = *pplSceneData.getCSMLayers();
+    const auto &shadows = *pplSceneData.getShadows();
+
+    const BuiltinCascadedShadowMap *result = nullptr;
+
+    { // find or create csm info
+        const BuiltinCascadedShadowMapKey key{&camera, mainLight};
+        auto iter = ppl.builtinCSMs.find(key);
+        if (iter != ppl.builtinCSMs.end()) {
+            result = &iter->second;
+        } else {
+            result = addBuiltinCSMInfo(ppl, pplSceneData, camera, mainLight, key, csmLayers);
+        }
+    }
+    CC_ENSURES(result);
+
+    // shadow not enabled
+    if (!shadows.isEnabled()) {
+        return nullptr;
+    }
+    // shadow type is planar
+    if (shadows.getType() == scene::ShadowType::PLANAR) {
+        return nullptr;
+    }
```

Comment:
I think it is better to check them as soon as possible.

Reply:
fixed.